File: 10-heap/00703-Kth_Largest_Element_In_A_Stream/00703-kth_largest_element_in_a_stream.go
// 703: Kth Largest Element In A Stream
// https://leetcode.com/problems/kth-largest-element-in-a-stream

package main

import (
	"container/heap"
	"fmt"
)

type Heap []int

func (h Heap) Len() int            { return len(h) }
func (h Heap) Less(i, j int) bool  { return h[i] < h[j] }
func (h Heap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *Heap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *Heap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// SOLUTION
type KthLargest struct {
	pq   *Heap
	size int
}

func Constructor(k int, nums []int) KthLargest {
	h := &Heap{}
	heap.Init(h)
	for i := 0; i < len(nums); i++ {
		heap.Push(h, nums[i])
	}
	for len(*h) > k {
		heap.Pop(h)
	}
	return KthLargest{h, k}
}

func (this *KthLargest) add(val int) int {
	if len(*this.pq) < this.size {
		heap.Push(this.pq, val)
	} else if val > (*this.pq)[0] {
		heap.Push(this.pq, val)
		heap.Pop(this.pq)
	}
	return (*this.pq)[0]
}

func main() {
	// INPUT
	k := 3
	nums := []int{4, 5, 8, 2}
	o := Constructor(k, nums)

	// OPERATIONS
	fmt.Println(o.add(3))
	fmt.Println(o.add(5))
	fmt.Println(o.add(10))
	fmt.Println(o.add(9))
	fmt.Println(o.add(4))
}

File: 10-heap/00703-Kth_Largest_Element_In_A_Stream/00703-kth_largest_element_in_a_stream.py
# 703: Kth Largest Element In A Stream
# https://leetcode.com/problems/kth-largest-element-in-a-stream


import heapq


class KthLargest:
    def __init__(self, k: int, nums: list[int]) -> None:
        self.pq = nums
        self.size = k
        
        heapq.heapify(self.pq)
        while len(self.pq) > k:
            heapq.heappop(self.pq)

    def add(self, val: int) -> int:
        if len(self.pq) < self.size:
            heapq.heappush(self.pq, val)
        elif val > self.pq[0]:
            heapq.heapreplace(self.pq, val)
        return self.pq[0]



if __name__ == "__main__":

    # INPUT
    k = 3
    nums = [4,5,8,2]
    o = KthLargest(k, nums)

    # OPERATIONS
    print(o.add(3))
    print(o.add(5))
    print(o.add(10))
    print(o.add(9))
    print(o.add(4))
